modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java

```diff
@@ -2271,11 +2279,28 @@ private String _getSegmentsAnonymousUserId() {
 				SegmentsWebKeys.SEGMENTS_ANONYMOUS_USER_ID));
 	}
 
-	private long[] _getSegmentsEntryIds() {
+	private long[] _getSegmentsEntryIds(AssetListEntry assetListEntry) {
+		List<AssetListEntrySegmentsEntryRel> assetListEntrySegmentsEntryRels =
+			_assetListEntrySegmentsEntryRelLocalService.
+				getAssetListEntrySegmentsEntryRels(
+					assetListEntry.getAssetListEntryId(), QueryUtil.ALL_POS,
+					QueryUtil.ALL_POS);
+
+		List<Long> segmentEntryIds = new ArrayList<>();
+
+		for (AssetListEntrySegmentsEntryRel assetListEntrySegmentsEntryRel :
+				assetListEntrySegmentsEntryRels) {
+
+			segmentEntryIds.add(
+				assetListEntrySegmentsEntryRel.getSegmentsEntryId());
+		}
+
 		return _segmentsEntryRetriever.getSegmentsEntryIds(
 			_themeDisplay.getScopeGroupId(), _themeDisplay.getUserId(),
 			_requestContextMapper.map(
-				PortalUtil.getHttpServletRequest(_portletRequest)));
+				_portal.getOriginalServletRequest(
+					_portal.getHttpServletRequest(_portletRequest))),
```

Comment:
Hi @cgoncas 
Thank you for suggesting a solution to this. I have tested this implementation and found that this could affect to the Widget Page the same way as the Blank Page. I am not sure about this behaviour because I found a condition at https://github.com/liferay/liferay-portal/blob/8618a3437b6fca76e6f909e8aad4fe0b17da0649/modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java#L149.
Could you please double-check?
Thank you very much.

Reply:
Hey @hongvo2308 !!

This is not a problem, the reason we run this code in the service pre action, is to avoid calculate the segments when they are not needed.

This calculation only affect the experiences, not the asset publisher.

Thanks!!

Reply:
Hi @cgoncas, 
Yes, I got it now.
Thank you for the explanation.